### PR TITLE
[1.9.x] Use cache change pool when DYNAMIC_RESERVE [7020]

### PR DESCRIPTION
**Pull request by IkerLuengo**
Adds a DYNAMIC_REUSABLE memory mode

This memory mode starts with an empty pool (much like DYNAMIC),
but the allocated slots for CacheChange are not freed, they are
added to the pool of free changes for future reuse insted.

### CODE REVIEW

File: src/cpp/rtps/history/CacheChangePool.cpp [hunk 1]
```diff
@@ -133,13 +136,40 @@ bool CacheChangePool::reserve_Cache(CacheChange_t** chan, uint32_t dataSize)
...
+                *chan = allocateSingle(dataSize); //Allocates a single, empty CacheChange, only if none free
+                if(*chan == nullptr)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                *chan = m_freeCaches.back();
+                m_freeCaches.erase(m_freeCaches.end()-1);
```
Comment (MiguelCompany) [inline src/cpp/rtps/history/CacheChangePool.cpp]: `*chan = m_freeCaches.pop_back()`

May be other parts of the code may also be modified

Comment (IkerLuengo) [reply src/cpp/rtps/history/CacheChangePool.cpp]: done and reviewed on the rest of the document

File: src/cpp/rtps/history/CacheChangePool.cpp [hunk 2]
```diff
@@ -118,8 +120,7 @@ bool CacheChangePool::reserve_Cache(CacheChange_t** chan, uint32_t dataSize)
                     return false;
                 }
             }
-            *chan = m_freeCaches.back();
-            m_freeCaches.erase(m_freeCaches.end()-1);
+            *chan = m_freeCaches.pop_back();
```
Comment (MiguelCompany) [inline src/cpp/rtps/history/CacheChangePool.cpp]:
```suggestion
            *chan = m_freeCaches.back();
            m_freeCaches.pop_back();
```
That should be the way

Comment (IkerLuengo) [reply src/cpp/rtps/history/CacheChangePool.cpp]: done. AND checked it compiles :S

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1044/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/866/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1042/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/337/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1061/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/882/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1059/)

File: src/cpp/rtps/history/CacheChangePool.cpp [hunk 3]
```diff
@@ -193,14 +202,31 @@ void CacheChangePool::release_Cache(CacheChange_t* ch)
...
+void CacheChangePool::return_cache_to_pool(CacheChange_t* ch)
+{
+    ch->kind = ALIVE;
+    ch->sequenceNumber.high = 0;
+    ch->sequenceNumber.low = 0;
+    ch->writerGUID = c_Guid_Unknown;
+    ch->serializedPayload.length = 0;
+    ch->serializedPayload.pos = 0;
+    for(uint8_t i=0;i<16;++i)
+        ch->instanceHandle.value[i] = 0;
```
Comment (MiguelCompany) [inline src/cpp/rtps/history/CacheChangePool.cpp]: I would use `memset` here

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1111/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/925/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1111/)

**Comment by MiguelCompany** [comment]
Failing tests are not related.

**Pull request merged**